json/src/fs2/data/json/internal/Renderer.scala

```diff
@@ -0,0 +1,174 @@
+    @tailrec
+    def loop(idx: Int): Unit =
+      if (idx < s.length) {
+        val c = s(idx)
+        (c: @switch) match {
+          case '\\' =>
+            builder.append("\\\\")
+          case '\b' =>
+            builder.append("\\b")
+          case '\f' =>
+            builder.append("\\f")
+          case '\n' =>
+            builder.append("\\n")
+          case '\r' =>
+            builder.append("\\r")
+          case '\t' =>
+            builder.append("\\t")
+          case _ =>
+            if (c > 127 || Character.isISOControl(c))
+              // escape non ascii or control characters
+              builder
+                .append("\\u")
+                .append((c >> 12) & 0x0f)
+                .append((c >> 8) & 0x0f)
+                .append((c >> 4) & 0x0f)
+                .append(c & 0x0f)
+            else
+              builder.append(c)
+        }
+        loop(idx + 1)
+      }
+
+    prefixValue()
+    builder.append('"')
+    loop(0)
+    builder.append('"')
+
+    if (key) {
+      newline = false
+      comma = false
+    } else {
+      newline = true
+      comma = level > 0
+    }
+  }
+
+  private def prefixValue(): Unit = {
+    if (comma)
+      builder.append(',')
+    if (pretty)
+      indentation(newline, level)
+  }
+
+  private def jsonValue(repr: String): Unit = {
+    prefixValue()
+    builder.append(repr)
+
+    newline = true
+    comma = level > 0
+  }
+
+  def +=(chunk: Chunk[Token]): Unit = {
+    if (resetOnChunk)
+      builder.setLength(0)
+    chunk.foreach {
+      case Token.StartObject =>
+        prefixValue()
+        builder.append('{')
+
+        level += 1
+        newline = true
+        comma = false
+      case Token.EndObject =>
+        if (pretty && comma)
+          indentation(newline, level - 1)
+
+        builder.append('}')
+
+        level -= 1
+        newline = true
+        comma = level > 0
+      case Token.StartArray =>
+        prefixValue()
+        builder.append('[')
+
+        level += 1
+        newline = true
+        comma = false
+      case Token.EndArray =>
+        if (pretty && comma)
+          indentation(newline, level - 1)
+
+        builder.append(']')
+
+        level -= 1
+        newline = true
+        comma = level > 0
+      case Token.Key(key) =>
+        jsonString(key, true)
+        builder.append(':')
+        if (pretty)
+          builder.append(' ')
+      case Token.TrueValue =>
+        jsonValue("true")
+      case Token.FalseValue =>
+        jsonValue("false")
+      case Token.NullValue =>
+        jsonValue("null")
+      case Token.NumberValue(repr) =>
+        jsonValue(repr)
+      case Token.StringValue(s) =>
+        jsonString(s, false)
+    }
+  }
+
+  def result: String = builder.result()
+
+}
+
+private[json] object Renderer {
+
+  def pipe[F[_]](pretty: Boolean, indent: String): Pipe[F, Token, String] = {
+    val builder = new Renderer(pretty, true, indent)
```

Comment:
I have the suspicion this isn't safe 🤔

If you acquire this pipe once and then use it in concurrently for several streams, I feel like these could interfere in the local state of the `Renderer`. To be sure, the renderer should probably be acquired effectfully or at least suspended:

```scala
in => Stream.suspend(Stream.pure(new Renderer(pretty, true, indent))).flatMap { renderer =>
  in.mapChunks { chunk =>
       renderer += chunk
       Chunk.singleton(renderer.result)
     }
}
```

Reply:
You are correct, mister, I'll fix this.

Reply:
I wonder if I can't just
```scala
    { in =>
      val renderer = new Renderer(pretty, true, indent)
      in.mapChunks { chunk =>
       renderer += chunk
       Chunk.singleton(renderer.result)
      }
    }
```

Reply:
Nevermind, of course I can’t…